### PR TITLE
Remove more `type_` methods in the JIT IR.

### PR DESCRIPTION
**Pull request by ltratt**
This is, I believe, the final push: there is now just `Method::_type` left in the JIT IR.

### CODE REVIEW

**Comment by ltratt** [comment]
Whoops, I pressed that on the wrong repo!

File: ykrt/src/compile/jitc_yk/jit_ir.rs [hunk 1]
```diff
@@ -811,24 +811,6 @@ impl Operand {
         }
     }
 
-    /// Returns the type of the operand.
```
Comment (vext01) [inline ykrt/src/compile/jitc_yk/jit_ir.rs:814]: What's the motivation for removing this? I don't mind it. It makes finding the types of things shorter.

Comment (ltratt) [reply ykrt/src/compile/jitc_yk/jit_ir.rs:814]: It's part of the general "make the user aware that the IR is a tree and you should get info from `Module`" push. Note that the methods removed in this PR either 1) weren't used 2) entirely duplicated existing `ty_idx` functions.

Comment (vext01) [reply ykrt/src/compile/jitc_yk/jit_ir.rs:814]: I'm kinda still luke warm on this.

It hurts my head more to ask "whats the type index of this, and what's the type for this type index" instead of "what the type of this".

A compromise would be to make the function simply do: " self.m.type_(op.ty_idx(self.m))" -- that big match isn't necessary.

Comment (vext01) [reply ykrt/src/compile/jitc_yk/jit_ir.rs:814]: > make the user aware that the IR is a tree

Dumb question: why does the API user need to care about this? Abstraction vs. implementation details, no?

File: ykrt/src/compile/jitc_yk/codegen/x86_64/mod.rs [hunk 2]
```diff
@@ -413,7 +419,7 @@ impl<'a> X64CodeGen<'a> {
             let op = &args[i];
             // We can type check the static args (but not varargs).
             debug_assert!(
-                i >= fty.num_args() || op.type_(self.m) == fty.arg_type(self.m, i),
+                i >= fty.num_args() || self.m.type_(op.ty_idx(self.m)) == fty.arg_type(self.m, i),
```
Comment (vext01) [inline ykrt/src/compile/jitc_yk/codegen/x86_64/mod.rs:422]: Should a follow up commit rename `type_` to `ty`, as per our new conventions?

File: ykrt/src/compile/jitc_yk/jit_ir.rs (hunk 1 above)
Comment (ltratt) [reply ykrt/src/compile/jitc_yk/jit_ir.rs:814]: Well we've done it everywhere else... I think it's better that we say "return indexes" consistently now. Or, if we want to have helper methods, to think carefully about why/when. Just having one or two left in feels like the worst of all worlds to me, I must admit.

File: ykrt/src/compile/jitc_yk/codegen/x86_64/mod.rs (hunk 2 above)
Comment (ltratt) [reply ykrt/src/compile/jitc_yk/codegen/x86_64/mod.rs:422]: Good point. Yes!

Comment (ltratt) [reply ykrt/src/compile/jitc_yk/codegen/x86_64/mod.rs:422]: Well, follow-up PR.

File: ykrt/src/compile/jitc_yk/jit_ir.rs (hunk 1 above)
Comment (vext01) [reply ykrt/src/compile/jitc_yk/jit_ir.rs:814]: OK.

Initially I'd aimed to make the API abstracted, instead of exposing the underlying data structures, but I guess you can't really do a full job of abstraction anyway.

Comment (ltratt) [reply ykrt/src/compile/jitc_yk/jit_ir.rs:814]: > Dumb question: why does the API user need to care about this? Abstraction vs. implementation details, no?

Good question! Normally I agree that you want to hide implementation details, but in this case we want to expose them, because it's *really* easy to treat the IR as if it's stored as a graph. That causes us to put lots of spaghetti-helpers everywhere that are extremely hard to follow. An extreme example of this was highlighted by https://github.com/ykjit/yk/pull/1116. Basically it's better, IMHO, to have the IR structs be as "dumb" as possible, making the user remember that `Module` actually stores all the interesting bits. This causes a bit more boilerplate but since it's very uniform, it actually causes less cognitive burden in the long run (IMHO).

File: ykrt/src/compile/jitc_yk/codegen/x86_64/mod.rs (hunk 2 above)
Comment (vext01) [reply ykrt/src/compile/jitc_yk/codegen/x86_64/mod.rs:422]: OK
